tools/nnyeah/nnyeah/Reworker.cs

```diff
@@ -0,0 +1,481 @@
+using System;
+using System.IO;
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+using Mono.Cecil.Rocks;
+using System.Linq;
+using System.Collections.Generic;
+
+namespace nnyeah {
+	public class Reworker {
+		Stream stm;
+		ModuleDefinition module;
+		TypeDefinition embeddedAttributeTypeDef;
+		TypeDefinition nativeIntegerAttributeTypeDef;
+		TypeReference nativeIntegerAttributeTypeRef;
+		TypeReference compilerGeneratedAttributeTypeRef;
+		TypeReference embeddedAttributeTypeRef;
+		TypeReference nintTypeReference;
+		TypeReference nuintTypeReference;
+
+		Dictionary<string, Transformation> methodSubs;
+		Dictionary<string, Transformation> fieldSubs;
+
+		public Reworker (Stream stm)
+		{
+			this.stm = stm;
+		}
+
+		public void Load ()
+		{
+			if (module != null)
+				return;
+			module = ModuleDefinition.ReadModule (stm);
+		}
+
+		public bool NeedsReworking ()
+		{
+			// simple prediate for seeing if there any references
+			// to types we need to care about.
+			// IntPtr is for future handling of types that
+			// descend from NObject and need to have the constructor
+			// changed to NHandle
+			return module.GetTypeReferences ().Any (
+				tr =>
+					tr.FullName == "System.nint" ||
+					tr.FullName == "System.nuint" ||
+					tr.FullName == "System.nfloat" ||
+					tr.FullName == "System.IntPtr"
+				);
+		}
+
+		void CheckModule ()
+		{
+			if (module == null)
+				throw new Exception ("Module is not loaded. Call Load first.");
+		}
+
+		public void Rework (Stream stm)
+		{
+			CheckModule ();
+
+			// get the types that we need to refer to later
+			FetchCompilerGeneratedAttributeTypeRef ();
+			AddEmbeddedAttributeIfNeeded ();
+			AddNativeIntegerAttributeIfNeeded ();
+			module.TryGetTypeReference ("System.nint", out nintTypeReference);
+			module.TryGetTypeReference ("System.nuint", out nuintTypeReference);
+
+			// load the substitutions
+			methodSubs = LoadMethodSubs ();
+			fieldSubs = LoadFieldSubs ();
+
+			foreach (var type in module.Types) {
+				ReworkType (type);
+			}
+
+			module.Write (stm);
+			stm.Flush ();
+		}
+
+		void ReworkType (TypeDefinition definition)
+		{
+			foreach (var field in definition.Fields) {
+				ReworkField (field);
+			}
+			foreach (var property in definition.Properties) {
+				ReworkProperty (property);
+			}
+			ReworkMethods (definition.Methods);
+			foreach (var @event in definition.Events) {
+				ReworkEvent (@event);
+			}
+			foreach (var innerType in definition.NestedTypes) {
+				ReworkType (innerType);
+			}
+		}
+
+		void ReworkMethods (IEnumerable<MethodDefinition> methods)
+		{
+			foreach (var method in methods) {
+				ReworkMethod (method);
+			}
+		}
+
+		void ReworkMethod (MethodDefinition method)
+		{
+			var nativeTypes = new List<bool> ();
+			var changed = false;
+			foreach (var parameter in method.Parameters) {
+				nativeTypes.Clear ();
+				changed = false;
```

Comment:
Why do we need to set change to false outside of the loop and not inside? This instance of change seems just to be needed inside the loop.